File: Homework/Hoang_Nguyen/lesson11/259_3SumSmaller.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumSmaller = function (nums, target) {
  let res = 0;
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length - 2; i++) {
    let start = i + 1;
    let end = nums.length - 1;
    while (start < end) {
      const sum = nums[i] + nums[start] + nums[end];
      if (sum < target) {
        res += end - start;
        start++;
      } else {
        end--;
      }
    }
  }
  return res;
};

File: Homework/Ngan_Luu/259-3sum-smaller.py
# https://leetcode.com/problems/3sum-smaller/
class Solution:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        nums.sort()
        res=0
        for i in range(0,len(nums)-2):
            start=i+1
            end=len(nums)-1
            while start<end:
                if nums[i]+nums[start]+nums[end]>=target:
                    end-=1
                else:
                    res+=end-start
                    start+=1
        return res